### DISCUSSION TITLE
How do I make an ANE respect elevated mode on Windows?

### DISCUSSION THREAD

**Discussion by manishpatel0404** [airsdk/Adobe-Runtime-Support#2295] [Q&A] [answered]
I have an AIR app on Windows 7, running in **elevated mode**. I also have an ANE coded in C that calls a WinAPI function `SetLocalTime()` which requires elevated privileges.

However, the `SetLocalTime()` function returns with an error, `A required privilege is not held by the client.` despite the app itself running in elevated mode with Administrator privileges.

The function works perfectly when I compile the same C code separately and run it as a Windows executable.

Is there a way to make the ANE run in elevated mode, if the AIR app is also run in the same mode?

Am I missing something, or is it not allowed by design, or is it a bug?

Look forward to some insight into this.

Thanks.

**Comment by ajwfrost** [comment, marked as answer]
Hi

I can't quite see why the function would fail if the app is running in elevated mode. The privileges should be applied to the process, so it's not going to depend on the various DLLs that are loaded in by the EXE file (including the AIR runtime and the ANEs)..

The only slight thing I wonder is if the library isn't code-signed, or maybe the fact it's not signed by the same certificate as the main binary is affecting things. We can have a bit of a play perhaps...

Just to confirm though -> this is a 'normal' ANE on Windows, where it's just a DLL that's being loaded in?
And have you checked what happens with the same binaries if you run them on Windows 10?

thanks

**Reply by manishpatel0404** [reply]
Yes, this is a normal ANE with a Windows-only implementation, through a DLL. The ANE is coded in C and compiled using `gcc` on MingW. However, this should not be a issue, because the same C code works when compiled as a Windows executable.

I have used the same certificate to sign all the ANEs and apps for my Company.

I have not tried to run the app on Windows 10, yet, because I am too used to Windows 7 Classic UI :) but will give it a shot. Should it matter, though?

Thanks.

**Reply by manishpatel0404** [reply]
@ajwfrost you were correct. If the app is running in elevated mode, the privileges are indeed applied to the process.

There was an error in my code. I missed out filling one field of the `SYSTEMTIME` struct, because of which `SetLocalTime()` returned with an error. It had nothing to do with special privileges.

Also, having / not having a signed library does not seem to matter. I tried both methods and both worked, once I fixed the above error :(

What really foxes me still, is that the C code compiled to a Windows executable still gives me the correct result, despite containing the error. This is actually what put me on the wrong track. Will definitely try to figure out why, once I get some rest.

Thanks.

**Reply by ajwfrost** [reply]
Okay that's good to know! Sometimes the error messages are a little odd :-)

Just replying to your earlier question about whether it mattered running on Windows 7 vs Windows 10 ... not really / I would hope not, but we generally use Windows 10 so I didn't want to start saying "it all works fine for us" if it was behaviour that was different between the different Windows editions!

thanks
